Documentation/ABI/JSON.md

```diff
@@ -0,0 +1,227 @@
+syntactically valid.
+
+### Common data types
+
+`<string>` and `<number>` are defined as in JSON. `<array:T>` represents an
+array (also defined as in JSON) whose elements all follow rule `<T>`.
+
+```
+<bool> ::= true | false ; as in JSON
+
+<source-location> ::= {
+  "fileID": <string>, ; the Swift file ID of the file
+  "line": <number>,
+  "column": <number>,
+}
+
+<timestamp> ::= {
+  "absolute": <number>, ; floating-point seconds since system-defined epoch
+  "since1970": <number>, ; floating-point seconds since 1970-01-01 00:00:00 UT
+}
+
+<version> ::= "version": 0 ; will be incremented as the format changes
+```
+
+<!--
+TODO: implement input/configuration
+
+### Configuration
+
+A single configuration is passed into the testing library prior to running any
+tests and, as the name suggests, configures the test run. The configuration is
+encoded as a single [JSON Lines](https://jsonlines.org) value.
+
+```
+<configuration-record> ::= {
+  <version>,
+  "kind": "configuration",
+  "payload": <configuration>
+}
+
+<configuration> ::= {
+  ["verbosity": <number>,] ; 0 is the default; higher means more verbose output
+                           ; while negative values mean quieter output.
+  ["filters": <array:test-filter>,] ; how to filter the tests in the test run
+  ["parallel": <bool>,] ; whether to enable parallel testing (on by default)
+  ; more TBD
+}
+
+<test-filter> ::= <test-filter-tag> | <test-filter-id>
+
+<test-filter-action> ::= "include" | "exclude"
+
+<test-filter-tag> ::= {
+  "action": <test-filter-action>,
+  "tags": <array:string>, ; the names of tags to include
+  "operator": <test-filter-tag-operator> ; how to combine the values in "tags"
+}
+
+<test-filter-tag-operator> ::= "any" | "all"
+
+<test-filter-id> ::= {
+  "action": <test-filter-action>,
+  "id": <test-id> ; the ID of the test to filter in/out
+}
+```
+-->
+
+### Streams
+
+A stream consists of a sequence of values encoded as [JSON Lines](https://jsonlines.org).
+A single instance of `<output-stream>` is defined per test process and can be
+accessed by passing `--experimental-event-stream-output` to the test executable
+created by `swift build --build-tests`.
+
+```
+<output-stream> ::= <output-record>\n | <output-record>\n <output-stream>
+```
+
+### Records
+
+Records represent the values produced on a stream. Each record is encoded on a
+single line and can be decoded independently of other lines. If a decoder
+encounters a record whose `"kind"` field is unrecognized, the decoder should
+ignore that line.
+
+```
+<output-record> ::= <metadata-record> | <test-record> | <event-record>
+
+<metadata-record> ::= {
+  <version>,
+  "kind": "metadata",
+  "payload": <metadata>
+}
+
+<test-record> ::= {
+  <version>,
+  "kind": "test",
+  "payload": <test>
+}
+
+<event-record> ::= {
+  <version>,
+  "kind": "event",
+  "payload": <event>
+}
+```
+
+### Metadata
+
+Metadata records are reserved for future use.
```

Comment:
This one feels quite speculative; if we need additional kinds of records, it seems like we can define them and extend this schema to describe them at that time, right? Rather than needing to explicitly reserve this now.

Reply:
Will delete for now. I think we had the idea that we wanted to front-load some additional data, but we can amend this schema with that if/when we implement it.